### ISSUE TITLE
tf.raw_ops.CropAndResizeGradImage crashes with rank-1 empty boxes tensor

### ISSUE THREAD

**Issue by yyds1233** [tensorflow/tensorflow#123397]
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

Yes

### Source

source

### TensorFlow version

tf.version.VERSION: 2.21.0-dev0+selfbuilt

### Custom code

Yes

### OS platform and distribution

Linux Ubuntu 22.04

### Python version

Python 3.10.12

### Bazel version

bazel-9.1.0

### GCC/compiler version

gcc (Ubuntu 11.4.0-1ubuntu1~22.04.2) 11.4.0

### Current behavior?

Calling tf.raw_ops.CropAndResizeGradImage with a rank-1 empty boxes tensor can crash the TensorFlow process instead of raising a clean, catchable TensorFlow exception.

The reproducer below uses only very small tensors:

grads shape:      [0, 1, 1, 1]
boxes shape:      [0]
box_ind shape:    [0]
image_size:       [2, 7, 7, 1]
T:                tf.float32
method:           "bilinear"

Here, boxes is an empty rank-1 tensor:

boxes = tf.constant([], dtype=tf.float32)

For CropAndResizeGradImage, boxes is expected to represent crop boxes, so an empty valid boxes tensor would normally have shape [0, 4]. A rank-1 empty tensor with shape [0] is malformed and should be rejected deterministically.

In my latest TensorFlow build, this input does not fail with a recoverable Python-level exception such as InvalidArgumentError. Instead, the process terminates with a fatal signal.

This can cause a denial-of-service condition when CropAndResizeGradImage receives malformed crop metadata derived from untrusted input. Invalid boxes shapes should be rejected before entering the native image-gradient kernel path.

### Standalone code to reproduce the issue

```shell
import tensorflow as tf

grads = tf.zeros([0, 1, 1, 1], dtype=tf.float32)
boxes = tf.constant([], dtype=tf.float32)
box_ind = tf.constant([], dtype=tf.int32)
image_size = tf.constant([2, 7, 7, 1], dtype=tf.int32)

y = tf.raw_ops.CropAndResizeGradImage(
    grads=grads,
    boxes=boxes,
    box_ind=box_ind,
    image_size=image_size,
    T=tf.float32,
    method="bilinear",
)

tf.reduce_sum(y).numpy()
```

### Relevant log output

```shell
F0000 00:00:1784211510.192957  420821 tensor_shape.cc:45] Check failed: NDIMS == dims() (2 vs. 1) Asking for tensor of 2 dimensions from a tensor of 1 dimensions
*** Check failure stack trace: ***
    @     0x7fe570f936aa  absl::lts_20250814::log_internal::LogMessage::SendToLog()
    @     0x7fe570f935ea  absl::lts_20250814::log_internal::LogMessage::Flush()
    @     0x7fe5708c7e8e  tensorflow::TensorShape::CheckDimsEqual()
    @     0x7fe57060c610  tensorflow::Tensor::tensor<>()
    @     0x7fe5415e634a  tensorflow::CropAndResizeGradImageOp<>::ComputeAsync()::{lambda()#1}::operator()()
    @     0x7fe5415ad036  tensorflow::(anonymous namespace)::RunIfBoxIndexIsValid<>()
    @     0x7fe5415e5961  tensorflow::CropAndResizeGradImageOp<>::ComputeAsync()
    @     0x7fe57055b2d0  tensorflow::AsyncOpKernel::Compute()
    @     0x7fe56f2c42c1  tensorflow::ThreadPoolDevice::Compute()
    @     0x7fe56f24492d  tensorflow::(anonymous namespace)::SingleThreadedExecutorImpl::Run()
    @     0x7fe56f1f564e  tensorflow::FunctionLibraryRuntimeImpl::RunSync()
    @     0x7fe56f20716c  tensorflow::ProcessFunctionLibraryRuntime::RunMultiDeviceSync()
    @     0x7fe56f2103c2  tensorflow::ProcessFunctionLibraryRuntime::RunSync()
    @     0x7fe55a99bb15  tensorflow::KernelAndDeviceFunc::Run()
    @     0x7fe55a917f69  tensorflow::EagerKernelExecute()
    @     0x7fe55a92794d  tensorflow::ExecuteNode::Run()
    @     0x7fe55b2b5abd  tensorflow::EagerExecutor::SyncExecute()
    @     0x7fe55a917343  tensorflow::(anonymous namespace)::EagerLocalExecute()
    @     0x7fe55a912d48  tensorflow::DoEagerExecute()
    @     0x7fe55a919087  tensorflow::EagerExecute()
    @     0x7fe55a20a587  tensorflow::EagerOperation::Execute()
    @     0x7fe55a99643a  tensorflow::CustomDeviceOpHandler::Execute()
    @     0x7fe555043f9f  TFE_Execute
    @     0x7fe51c7d83f2  TFE_Py_FastPathExecute_C()
    @     0x7fe47666699a  pybind11::detail::argument_loader<>::call<>()
    @     0x7fe47666682c  pybind11::cpp_function::initialize<>()::{lambda()#1}::__invoke()
    @     0x7fe476624521  pybind11::cpp_function::dispatcher()
    @     0x55fa03107962  (unknown)
Aborted (core dumped)
```

**Comment by Maanvi212006** [comment]
Hi @yyds1233 @Venkat6871 @Kayyuri , I'd like to work on this issue — could you assign it to me?

**Approach:** The crash happens because `CropAndResizeGradImageOp::ComputeAsync` assumes `boxes` has rank 2 (shape `[num_boxes, 4]`) and calls `tensor<T, 2>()` on it without validating rank first, so a rank-1 `boxes` tensor trips the internal `CHECK failed: NDIMS == dims()` in `tensor_shape.cc` and aborts the process.

I'll add explicit shape validation early in `ComputeAsync` (before the async kernel path runs) that checks `boxes.dims() == 2` and `boxes.dim_size(1) == 4`, returning `errors::InvalidArgument` via `OP_REQUIRES_ASYNC` instead of letting it hit the native path. I'll also check whether `box_ind` needs a similar guard, then add a Python test reproducing this exact case and confirming it now raises a catchable `InvalidArgumentError`.

**Comment by Kayyuri** [comment]
Hi @Maanvi212006, Thank you for your interest in TensorFlow.
Since this is related to open-source TensorFlow, so no individual contributors need to be assigned. TensorFlow welcomes contributions from everyone. Anyone interested in contributing can follow the provided guidelines, which I’ve attached [here](https://www.tensorflow.org/community/contribute) for your reference.
Thanks!
